File: frontend/src/lib/environment.ts
/**
 * Which deployment is this? — runtime environment identity for the UI.
 *
 * AIPLA runs three near-identical deployments whose only visible difference is
 * an opaque Cloud Run hostname (`aipla-v01-frontend-wgwhd7mspa` vs `…-y2bmxayxca`
 * vs `…-6vwz657g3a`). On 2026-08-04 a teacher minted group codes on dev and
 * typed them into test for two hours, getting a 401 on every join, because
 * group codes are Firestore documents and Firestore is per-project. The URL is
 * not a usable signal for a human, so the UI has to say it in words.
 *
 * The answer comes from the BACKEND at request time, not from a
 * `NEXT_PUBLIC_*` build arg: prod runs the byte-identical image that test
 * built (build-once artifact promotion), so anything compiled in would label
 * prod as "test". See backend/config/environment.py.
 */

export type EnvironmentName = "dev" | "test" | "prod" | "local" | "unknown";

export interface EnvironmentInfo {
  env: EnvironmentName;
  projectId: string | null;
  version: string | null;
}

const KNOWN: readonly EnvironmentName[] = [
  "dev",
  "test",
  "prod",
  "local",
  "unknown",
];

/**
 * In-flight/settled fetch, shared across every component that asks. The banner
 * mounts on every page; without this each remount would re-request.
 */
let cached: Promise<EnvironmentInfo | null> | null = null;

async function requestEnvironment(): Promise<EnvironmentInfo | null> {
  try {
    // Public endpoint — deliberately no auth header. The student join page is
    // the surface that most needs the label and nobody is signed in there.
    const res = await fetch("/api/proxy/api/environment", { cache: "no-store" });
    if (!res.ok) return null;
    const body = (await res.json()) as Partial<EnvironmentInfo>;
    const env = KNOWN.includes(body.env as EnvironmentName)
      ? (body.env as EnvironmentName)
      : "unknown";
    return {
      env,
      projectId: body.projectId ?? null,
      version: body.version ?? null,
    };
  } catch {
    // Unreachable backend is not "this is production" — the caller treats
    // null as "don't claim to know", and renders nothing.
    return null;
  }
}

export function fetchEnvironment(): Promise<EnvironmentInfo | null> {
  cached ??= requestEnvironment();
  return cached;
}

/** Test seam — drops the shared cache so each test starts cold. */
export function resetEnvironmentCache(): void {
  cached = null;
}

/**
 * Should the environment be called out to the user?
 *
 * Prod is the one that needs no banner: it's the real thing, and a permanent
 * strip on the site teachers actually use is noise. Everything else — including
 * `unknown` — gets labelled. `local` is excluded because LocalModeBanner
 * already says it, louder.
 */
export function shouldAnnounceEnvironment(env: EnvironmentName): boolean {
  return env === "dev" || env === "test" || env === "unknown";
}

export interface EnvironmentLabel {
  /** Short all-caps tag, e.g. "TEST". */
  tag: string;
  /** One line, Danish (student-facing surfaces are bilingual). */
  da: string;
  /** The same line in English. */
  en: string;
}

export function environmentLabel(env: EnvironmentName): EnvironmentLabel {
  switch (env) {
    case "test":
      return {
        tag: "TEST",
        da: "Testmiljø til afprøvning — koder og klasser her virker kun på denne adresse.",
        en: "Test environment — codes and classes here work only on this address.",
      };
    case "dev":
      return {
        tag: "DEV",
        da: "Udviklingsmiljø — under opbygning. Data kan blive slettet uden varsel.",
        en: "Development environment — work in progress. Data may be wiped without warning.",
      };
    // prod and local never reach the banner (shouldAnnounceEnvironment), but
    // they DO reach the wrong-site hint on the join page, which names whichever
    // site the student is on. A "prod" that fell through to the unknown label
    // would tell a student on the real site that it is unrecognised.
    case "prod":
      return {
        tag: "AIPLA",
        da: "Det rigtige AIPLA — det miljø, undervisningen kører på.",
        en: "The live AIPLA site — the environment teaching runs on.",
      };
    case "local":
      return {
        tag: "LOCAL",
        da: "Lokal udgave på din egen maskine.",
        en: "Local build running on your own machine.",
      };
    case "unknown":
    default:
      return {
        tag: "UKENDT / UNKNOWN",
        da: "Ukendt miljø — kontrollér adressen med din underviser, før du bruger det.",
        en: "Unrecognised environment — check the address with your teacher before using it.",
      };
  }
}

File: backend/config/environment.py
"""Which deployment am I? — runtime environment identity.

Resolved at REQUEST time from env vars, never baked into the image. AIPLA
promotes ONE built artifact dev -> test -> prod (build-once artifact
promotion, see docs/design/aipla/v1.0.0-pilot/build-once-artifact-promotion.md),
so anything derived at BUILD time is wrong the moment the image is copied:
prod would introduce itself as "test" because that is where its bytes were
built. The trustworthy signals are the env vars Terraform sets per Cloud Run
service — ``GOOGLE_CLOUD_PROJECT`` is set on every environment's sidecar and
differs per project, so it works today with no infra change.

Why this exists at all: on 2026-08-04 a teacher minted group codes on dev and
typed them into the test deployment for ~2 hours, getting 401 on every join
(codes are Firestore documents, and Firestore is per-project). The three
``aipla-v01-frontend-*.run.app`` URLs are indistinguishable at a glance, so the
UI has to say which environment it is. This module is where the answer comes
from — see ``frontend/src/components/EnvironmentBanner.tsx`` for the surface.
"""

from __future__ import annotations

import os
import re

# ``aipla-dev-2026`` / ``aipla-test-2026`` / ``aipla-prod-2026``. The trailing
# year is part of the naming convention, not a version — match it loosely so a
# future ``aipla-prod-2027`` still resolves.
_PROJECT_PATTERN = re.compile(r"^aipla-(dev|test|prod)-\d{4}$")

#: Every name ``environment_name()`` can return. "unknown" is a real answer,
#: not an error case: an unrecognised deployment must still be LABELLED as
#: unrecognised rather than silently passing for production.
KNOWN_ENVIRONMENTS = ("dev", "test", "prod", "local", "unknown")


def environment_name() -> str:
    """Resolve this deployment's environment name.

    Order of trust:
      1. ``AIPLA_ENV`` — explicit override, for deployments whose project id
         doesn't follow the convention. Ignored if it isn't a known name.
      2. ``LOCAL_MODE`` — a laptop, whatever the project id says.
      3. ``GOOGLE_CLOUD_PROJECT`` matching ``aipla-<env>-<year>``.
      4. ``"unknown"``.
    """
    override = (os.getenv("AIPLA_ENV") or "").strip().lower()
    if override in KNOWN_ENVIRONMENTS:
        return override

    # Local import: config.local_mode reads env vars at call time too, and
    # importing it at module scope pulls its dependency graph into anything
    # that only wants the environment name.
    from config.local_mode import is_local_mode

    if is_local_mode():
        return "local"

    project = (os.getenv("GOOGLE_CLOUD_PROJECT") or os.getenv("GCP_PROJECT") or "").strip()
    match = _PROJECT_PATTERN.match(project)
    if match:
        return match.group(1)

    return "unknown"


def environment_info() -> dict:
    """The payload behind ``GET /api/environment``.

    ``version`` is the release tag Terraform stamps on the service
    (``APP_VERSION``), so a support conversation can pin down exactly what a
    teacher is looking at — "test, v0.1.5" — from one glance at the banner.
    """
    return {
        "env": environment_name(),
        "projectId": (os.getenv("GOOGLE_CLOUD_PROJECT") or os.getenv("GCP_PROJECT") or "").strip() or None,
        "version": (os.getenv("APP_VERSION") or "").strip() or None,
    }


__all__ = ["KNOWN_ENVIRONMENTS", "environment_info", "environment_name"]
